### PR TITLE
Add auto-labelling to pull requests

### PR DESCRIPTION
**Pull request by nilason**
This adds auto-labelling to pull requests (related to https://github.com/OSGeo/grass/issues/526).

It works with GitHub actions and add labels based on the PR's modified/added/deleted files location in repository.

This is a proposal, so if you find this approach appealing at all, I'm happy to receive suggestions for modifications.

Present proposal requires addition of some new labels.

The rules are stated in labeller.yml and as an example any modification in lib or include directory would add a `lib` label to the PR.

### CODE REVIEW

File: .github/labeler.yml
Comment (landam) [inline .github/labeler.yml:1]: This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

File: .github/labeler.yml
Comment (landam) [inline .github/labeler.yml:5]: This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

File: .github/labeler.yml
Comment (landam) [inline .github/labeler.yml:31]: This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

File: .github/labeler.yml
Comment (landam) [inline .github/labeler.yml:34]: We have already a label called `manual`. Either use `manual` or renamed existing label to `docs`.

File: .github/labeler.yml
Comment (landam) [inline .github/labeler.yml:37]: This label doesn't exist in GitHub yet, see also related discussion about labels at, https://trac.osgeo.org/grass/wiki/GitMigration#Migrationoftracissues

File: .github/labeler.yml
Comment (wenzeslaus) [reply .github/labeler.yml:34]: We were prefixing some commits with `docs:` and `doc:` in the past. `docs` is often used in Git repos for the documentation directory. `docs` might be broader than `manual`, so perhaps more useful as it can be applied to other things than just `doc` dir.

**Comment by wenzeslaus** [comment]
I think this would be useful to have or at least try out in action even if it is limited e.g. to the currently existing labels (if we want to keep the number of labels low).

**Comment by neteler** [comment]
> I think this would be useful to have or at least try out in action even if it is limited e.g. to the currently existing labels (if we want to keep the number of labels low).

We may try, otherwise we'll never find out...

**Comment by neteler** [comment]
@nilason does this PR need to be updated since GH Actions are now in place?

**Comment by nilason** [comment]
> @nilason does this PR need to be updated since GH Actions are now in place?

No, this is a separate feature, should work as is.

File: .github/labeler.yml
Comment (neteler) [reply .github/labeler.yml:31]: `docker`is meanwhile there: https://github.com/OSGeo/grass/labels

File: .github/labeler.yml
Comment (neteler) [reply .github/labeler.yml:34]: Shall we rename the current label https://github.com/OSGeo/grass/labels/manual to docs? (fine for me)

File: .github/labeler.yml
Comment (neteler) [reply .github/labeler.yml:37]: Is `locale` really clear to everybody or better a different word?

File: .github/labeler.yml
Comment (neteler) [reply .github/labeler.yml:1]: `lib` or `libraries`? I'd prefer the latter for being less cryptic.

File: .github/labeler.yml
Comment (neteler) [reply .github/labeler.yml:5]: It's `core` more like `modules`? To distinguish it from `lib(s)` or `libraries`?

**Review by neteler — DISMISSED** [review]
Some comments added inline...

File: .github/labeler.yml
Comment (nilason) [reply .github/labeler.yml:5]: I was thinking of naming this modules, it might be a better choice.

File: .github/labeler.yml
Comment (nilason) [reply .github/labeler.yml:1]: I chose `lib` as it's shorter, but `libraries` is good for me too.

File: .github/labeler.yml
Comment (nilason) [reply .github/labeler.yml:37]: not absolutely necessary to have this label at all...

**Comment by nilason** [comment]
Rebased and updated to better reflect current list of labels.

The labels `libraries` and `module` (or alternative label)  need to be created for this to work.

I commented out `docs` and `manual`, for the time being.

**Comment by neteler** [comment]
PR rebased, but it needs more permissions:

```
Run actions/labeler@v4
...
Warning: The action requires write permission to add labels to pull requests. For more information please refer to the action documentation: https://github.com/actions/labeler#permissions
Error: Resource not accessible by integration
```

https://github.com/actions/labeler#permissions

**Comment by nilason** [comment]
> PR rebased, but it needs more permissions:
> 
> ```
> Run actions/labeler@v4
> ...
> Warning: The action requires write permission to add labels to pull requests. For more information please refer to the action documentation: https://github.com/actions/labeler#permissions
> Error: Resource not accessible by integration
> ```
> 
> https://github.com/actions/labeler#permissions

Thanks!

Updated, works now on a private test repo of mine. (Looks like it have to be merged before kicking in).

File: .github/labeler.yml
Comment (wenzeslaus) [inline .github/labeler.yml:4]: We don't have this label, no? We do have C, C++ and Python which are used below.

File: .github/labeler.yml
Comment (wenzeslaus) [inline .github/labeler.yml:10]: I can't find this label either. Maybe worth adding it. (?)

We have vector, raster, temporal, and database.

File: .github/labeler.yml
Comment (wenzeslaus) [inline .github/labeler.yml:46]: We have manual and HTML.

File: .github/labeler.yml
Comment (wenzeslaus) [inline .github/labeler.yml:38]: Hm, docker or Linux label. I don't know.

File: .github/labeler.yml
Comment (wenzeslaus) [inline .github/labeler.yml:52]: Add `.travis`.

File: .github/labeler.yml
Comment (wenzeslaus) [inline .github/labeler.yml:25]: Remove commented-out code, but you can include comment on what is included and what is not.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]

**Comment by wenzeslaus** [comment]
> Updated, works now on a private test repo of mine.

I don't see any permissions in the YAML file, but I see them in the example in doc linked by @neteler. Any insights? What I'm asking is basically, do you expect the workflow to successfully run after merging it with the current configuration in terms of permissions?

> (Looks like it have to be merged before kicking in).

That's definitively the case.

File: .github/labeler.yml
Comment (nilason) [reply .github/labeler.yml:4]: I believe my initial intention with `libraries` and `module`, was to be able to highlight evtl. changes to core vs. modules. Changes to libraries, being base to all, perhaps need particular attention.

Comment (wenzeslaus) [reply .github/labeler.yml:4]: I think you need to create the labels first. I think they might be useful, so why not, especially if we have automation to help us label.

File: .github/labeler.yml
Comment (nilason) [reply .github/labeler.yml:46]: Perhaps we can use the (existing) label `manual` for documentation in a broad sense like:
 
```
manual:
  - man/**/*
  - doc/**/*
```

File: .github/labeler.yml
Comment (nilason) [reply .github/labeler.yml:10]: Just added support to generate vector, raster, temporal, and database labels too.

File: .github/labeler.yml
Comment (wenzeslaus) [reply .github/labeler.yml:54]: Or rename manual to documentation or docs. That would make sense to me.

Comment (nilason) [reply .github/labeler.yml:54]: `docs` is short and concise! See also https://github.com/OSGeo/grass/pull/531#discussion_r422619437 and below...

Comment (nilason) [reply .github/labeler.yml]: updated to use `docs`

Comment (wenzeslaus) [reply .github/labeler.yml]: Sounds good. I agree with my 2020 self.

File: .github/labeler.yml
Comment (neteler) [reply .github/labeler.yml:46]: rpm could probably be skipped as it has never (?) received any reports in the past 10 years.

File: .github/workflows/label.yml
Comment (echoix) [inline .github/workflows/label.yml:17]:
```suggestion
      - uses: actions/labeler@v5
```
Latest action is now v5

File: .github/workflows/label.yml
Comment (echoix) [inline .github/workflows/label.yml:11]:
```suggestion
on:
  - pull_request_target
```

File: .github/workflows/label.yml
Comment (echoix) [inline .github/workflows/label.yml:14]:
```suggestion
  labeler:
    permissions:
      contents: read
      pull-requests: write
```

As the example of the action's docs.

File: .github/workflows/label.yml
Comment (echoix) [inline .github/workflows/label.yml:12]:
```suggestion
permissions: {}

```

GitHub action permissions has evolved since, and the newest best practice is to specify at the workflow level the lowest permissions needed, and at the job level the minimum permissions needed. These permissions declare what the `GITHUB_TOKEN` is granted to do.


The advantage of setting the least at the workflow level is that further changes to the workflow by adding a job doesn't unintentionally give more rights that was previously planned out, but overlooked afterwards.


Declaring the permissions are quite easier to understand the config when looking at the workflow code, instead of a web interface that changes over time.

File: .github/workflows/label.yml
Comment (echoix) [inline .github/workflows/label.yml:19]:
```suggestion
```
Not needed, it implicitly uses the GitHub token

Comment (echoix) [inline .github/workflows/label.yml:19]:
```suggestion
        with:
          sync-labels: true
```

Optionally, I think we could use the sync-labels option, and if it bothers us too much we (probably me, the first impacted), could remove it afterwards. 
https://github.com/actions/labeler#inputs

**Review by echoix — CHANGES_REQUESTED** [review]
I left some new suggestions to have this PR up to date with changes with the action used. I would really like this PR to come through soon, it would be quite useful for tidying up the backlog (that I started doing)

**Comment by echoix** [comment]
@neteler Just so you know, in this section it mentions that the action might fail one day at the time of a major version update of the action that would introduce breaking changes in the config, so that we might need to force it through while using the `pull_request_target` event type. That event type is really the most appropriate here however. They explain the workaround if ever needed.
https://github.com/actions/labeler#updating-major-version-of-the-labeler

**Comment by echoix** [comment]
And for the different labels, and the old discussions about various small details, I think that it's overall quite adequate, docs is a good idea, since having details on html, css, JavaScript might be overkill.

However, if is possible to make a distinction between docs, the content, and docs, the system/infrastructure around it, that might be a better solution than having things like "manual".

Docs-only changes should be able to be merged more quickly without the same amount of programming-level code review. My stance is that it is already both hard to have contributors in open source and hard to have good docs in open source, so let's not leave out the contributions we could have. Thus, when a PR would have docs (the content) only changes, if it's clearer and no mistakes, it could be merged expeditiously. Docs, as the infrastructure, could be like any other topic, and I think there is work on changing from manually creating HTML to markdown, so these extra categories will change in the mid-future.


In conclusion, this PR, as a tool for us, should be fine tuned after usage, and it's our (changing) usage that dictates its requirements. We can't be that wrong with this. Our irritants will show us what to improve.

File: .github/labeler.yml
Comment (echoix) [inline .github/labeler.yml:1]: I think that there might be missing a RFC filter, as they are being committed in the repo now, and should concern different members.

Comment (wenzeslaus) [reply .github/labeler.yml:1]: I agree. The label is https://github.com/OSGeo/grass/labels/RFC

**Comment by echoix** [comment]
Do I have the blessing to apply the suggestions I think are needed to unblock this PR?

**Comment by nilason** [comment]
> Do I have the blessing to apply the suggestions I think are needed to unblock this PR?

I'm sitting with it at this moment.

**Comment by nilason** [comment]
Thanks @echoix for the suggestions, added them with new commit.

**Comment by nilason** [comment]
May I suggest to merge this ASAP, this workflow may be fine-tuned or even dropped is not good enough.

**Comment by echoix** [comment]
I was about to finish approving it and go for merging. Merging main into the PR branch wasn't explicitly needed. We have enabled the option to always show it, so it was a suggestion rather than a warning, and it helps to retrigger CI for older PRs, without having to close/open the PR (which would run the same outdated CI), and rather pull in some new code that might include CI changes.

We'll see in 1h30+ if we can merge it then..

**Review by echoix — APPROVED** [review]

**Pull request merged**
